src/rate_limiter.rs

```diff
@@ -0,0 +1,247 @@
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::time::{Duration, Instant};
+
+use crate::TokenBucket;
+
+pub struct RateLimiter<'a, K> {
+    buckets: HashMap<K, TokenBucket<'a>>,
+    clock: &'a (dyn Fn() -> Instant + Sync),
+}
+
+impl<'a, K: Eq + Hash> RateLimiter<'a, K> {
+    pub fn new() -> Self {
+        Self::with_timer(&Instant::now)
+    }
+
+    fn with_timer(clock: &'a (dyn Fn() -> Instant + Sync)) -> Self {
+        Self {
+            buckets: HashMap::new(),
+            clock,
+        }
+    }
+
+    pub fn set_limit(&mut self, key: K, limit: usize, interval: Duration) {
```

Comment:
Let's merge this as is, but as a follow up think about moving this into a separate "builder" struct? My thinking is that we want to avoid acquiring a (global, per RateLimiter instance) lock on every call to `consume()`, and that would be one way to achieve by restricting what users can do via the struct API.

Reply:
> My thinking is that we want to avoid acquiring a (global, per RateLimiter instance) lock on every call to consume()

Are you talking about hypothetical lock to make `set_limit` thread-safe or something else?

Reply:
Yeah, making `RateLimiter` implement `Sync`, so that we can call `set_limit()` from multiple threads.

Reply:
Right. I think most people will want to configure rate-limits on start anyway, so having a separate builder makes most sense to me. If someone wants to modify the values later, they can always have a `RateLimiter` instance behind an `Arc` and atomically swap a pointer to a new instance.

Reply:
> You'd have to either require passing all limits before the RateLimiter is usable (IIRC this is what @malor is proposing above), making the hash map immutable and thus thread-safe or you'll have to update the API to expose individual TokenBuckets to avoid a global lock on every consume.

I believe we have discussed `.set_limit()` not being a thread-safe on the stream earlier today and came to the conclusion to implement the builder pattern.

> you'll have to update the API to expose individual TokenBuckets to avoid a global lock on every consume.

Which breaks the point of introducing the rate-limiter structure in the first place. So I don't think this an option. Those who don't want to pay for the rate-limiter abstraction can always construct and manage token buckets directly. So I think we're down to the builder pattern option.

> The only way to achieve reasonable performance with the current API (while making the implementation thread-safe) would be introducing thread-local caches of the hash-map.

I'm not sure I'm following your suggestion. I mean, wouldn't we have the exactly same issue on cache misses?

Reply:
> If someone wants to modify the values later, they can always have a RateLimiter instance behind an Arc and atomically swap a pointer to a new instance.

But that's going to be an instance with a fresh state. I'm not sure how useful it's going to be if you only want to add a new bucket without resetting others.

Reply:
> I'm not sure I'm following your suggestion. I mean, wouldn't we have the exactly same issue on cache misses?

Cache miss (or rather detection that value in cache is stale) will happen within consume once per thread after and update was called. With an assumption that updates are rather rare (compare to consume()) this is a reasonable trade off.

> behind an Arc and atomically swap a pointer to a new instance.

hehe, this is a common misconception about shared_ptrs in C++ and Arc. Even though Arc is safe to copy/read from multiple threads, it's actually not safe to assign/swap. 

Most strategies to deal with it usually optimize for reducing the overhead per read on the fast path while potentially slightly increasing the cost of write (e.g. a versioned thread local cache is an example of such strategy). I just found some crate that tries to implement and atomic Arc, optimized for fast reads - https://docs.rs/arc-swap/latest/arc_swap/. From quick read of their documentation (https://docs.rs/arc-swap/latest/arc_swap/docs/performance/index.html and https://docs.rs/arc-swap/latest/arc_swap/docs/limitations/index.html) their approach sounds similar to hazard pointer implementations.

Reply:
> But that's going to be an instance with a fresh state.

Main state would be (RWLock<Arc<HashMap<K, Arc<TokenBucket>>>>, atomic<Version>)
Cache is ThreadLocal<(Arc<HashMap<K, Arc<TokenBucket>>>, Version)>

consume():
1. Compares thread-local version with global version, if good skips to 3
2. Grabs the read-local and copies Arc into the thread-local state. The critical section is super small.
3. Uses HashMap from the thread-local cache

set_limit():
1. Acquire the write lock
2. Create a copy of a HashMap (copying existing TokenBucket Arcs, not TokenBuckets themselves)
3. Replace the global HashMap 
4. Bump the version
5. Release the lock

Note that this approach has a somewhat heavy critical section for the write path (copying the hash map). That can be further reduced by using an upgrade lock during step 2 and only getting the write lock for step 3 (this is not supported by Rust's RWLock but is normally present in locking libraries - e.g. https://github.com/facebook/folly/blob/main/folly/docs/Synchronized.md#intro-to-upgrade-mutexes. This can be simulated by an additional lock taken for upgrades). Alternatively some form of compare-and-swap can be implemented using version to detect that another write was performed (and potentially repeating 2-3 steps in a loop if such collisions happens), but this is strictly worse than an upgrade lock.

Reply:
> Cache miss (or rather detection that value in cache is stale) will happen within consume once per thread after and update was called. With an assumption that updates are rather rare (compare to consume()) this is a reasonable trade off.

Why do you call it "stale detection"? Maybe we're talkinig about different things. As I understand, your suggestion is to come maintain one main HashMap, and one HashMap for each thread (thread-local). Updated to the main HashMap is protected by mutex. Each thread uses its own thread-local hashmap without a lock to find a tokebucket instance to consume from. Whenever the key is missing in the thread-local map, the thread needs to acquire main HashMap mutex and copy the value from the main map to the thread-local map. Is that what you're suggesting?

Reply:
> > But that's going to be an instance with a fresh state.
>
> Main state would be (RWLock<Arc<HashMap<K, Arc>>>, atomic)
> Cache is ThreadLocal<(Arc<HashMap<K, Arc>>, Version)>

I think you have misquoted me. :) I have replied to @malor and his idea to atomicly replace rate-limiter instances.

Speaking about your suggestion, I don't think it's reasonable to introduce such over complication. I don't see how it's better than double-checked locking that don't require maintaining thread-local caches.

Reply:
> hehe, this is a common misconception about shared_ptrs in C++ and Arc. Even though Arc is safe to copy/read from multiple threads, it's actually not safe to assign/swap.

TIL. Somehow I expected both to work similarly to https://clojuredocs.org/clojure.core/swap! . Thank you for pointing this out! :+1: I'll check out ArcSwap.

Reply:
> I think you have misquoted me. :) I have replied to @malor and his idea to atomicly replace rate-limiter instances.

I know that it was a reply to @malor :) I was mostly explaining that you don't need to start from a fresh state if your HashMap stores Arc<TokenBuckets> rather than TokenBuckets.

>  I don't see how it's better than double-checked locking that don't require maintaining thread-local caches.

I'm not sure I understand what you mean by double-checked locking :(

I think I described multiple ideas in my comments. But the primary reason to use thread-local caches is to avoid any global locking (or even any global atomic) on the RateLimiter consume critical path.

I see 3 reasonable choices depending on what you want to get:
1. RateLimiter doesn't support adding keys after construction. Then RateLimiter doesn't need locking.
2. RateLimiter supports adding keys after construction and uses a global lock. In this case TokenBucket itself doesn't need to be thread safe, since the global lock will be limiting the performance.
3. RateLimiter supports adding keys and some form of thread-local strategy is used to avoid a global lock. Then each token buck still has to be thread-safe.

If you want to go for "not overcomplicated" you have to pick 1 or 2 :) If you are picking 2 you should drop thread-safety from the TokenBucket.

Reply:
Given that # of `consume()` calls is likely to be >> # of `set_limit()` calls, I wonder how a persistent [HashMap](https://docs.rs/im/latest/im/struct.HashMap.html) like `ArcSwap<im::HashMap<K, TokenBucket>>` would perform in this case:

- `consume()` only needs to load the value of `ArcSwap`'s atomic, similarly to the `version` atomic you had in the previous comments. It does _not_ contend with concurrent `consume()` or `set_limit()` calls
- `set_limit()` _does_ contend only with other concurrent `set_limit()` calls: it might need multiple retries to successfully update the hashmap's state, but those attempts are relatively cheap because the persistent data structures uses structural sharin

Reply:
> consume() only needs to load and dereference the value of ArcSwap's atomic pointer, similarly to the version atomic you had in the previous comments. It does not contend with concurrent consume() or set_limit() calls

You are making an assumption that such "dereference" is a regular dereference and that's incorrect. In practice once you have to support some form of concurrent updates such "deference" either needs to depends on some thread-local caches or it will involve some atomic CAS/increment operations. 

See https://github.com/vtyulb/AtomicSharedPtr or https://github.com/anthonywilliams/atomic_shared_ptr or https://github.com/facebook/folly/blob/main/folly/concurrency/AtomicSharedPtr.h for examples of lock-free implementations. The key idea is that it needs to somehow capture a reference and increment a ref-count in a single atomic operation (usually CAS). This can be done either via 128-bit CAS or some hacks that rely on packing a pointer and a ref-count into 64 bits.

Reply:
Ah, right: there is no guarantee that the referenced object will live long enough, unless we are holding an actual `Arc` that increments the reference counter :(